### PR TITLE
CORE-15179 add caching to claim handler

### PR DESCRIPTION
**Pull request by owenstanford**
This change adds caching to the token claim handler code to minimise calls to the DB.
The query that selects tokens is now configurable.
Some refactoring was made to improve some areas of the code.

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -46,7 +46,7 @@ bouncycastleVersion=1.73
 # Corda API libs revision (change in 4th digit indicates a breaking change)
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
 #cordaApiVersion=5.1.0.xxx-SNAPSHOT
-cordaApiVersion=5.1.0.15-beta+
+cordaApiVersion=5.1.0.15-alpha-1692633563033
```
Comment (filipesoliveira) [inline gradle.properties]: THIS HAS TO BE UPDATED BEFORE MERGING THE PR!

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/entities/TokenPoolCacheImpl.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package net.corda.ledger.utxo.token.cache.entities
+
+class TokenPoolCacheImpl : TokenPoolCache {
+
+    private val cache = mutableMapOf<TokenPoolKey, TokenCache>()
```
Comment (josephzunigadaly) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/entities/TokenPoolCacheImpl.kt]: Do we need a concurrent map? Is there any risk of concurrent access from the users of this class?

Comment (filipesoliveira) [reply components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/entities/TokenPoolCacheImpl.kt]: I believe the concurrent map is not needed since all HTTP requests are processed sequentially and there is no worker thread populating the cache in parallel.

File: components/ledger/ledger-utxo-token-cache/src/test/kotlin/net/corda/ledger/utxo/token/cache/impl/services/TokenCacheEventProcessorTest.kt [hunk 3]
```diff
@@ -30,6 +31,7 @@ class TokenCacheEventProcessorTest {
     private val tokenCacheEventHandlerMap = mutableMapOf<Class<*>, TokenEventHandler<in TokenEvent>>()
     private val event = FakeTokenEvent()
     private val tokenCache = mock<TokenCache>()
+    private val tokenPoolCache = mock<TokenPoolCache>()
```
Comment (josephzunigadaly) [inline components/ledger/ledger-utxo-token-cache/src/test/kotlin/net/corda/ledger/utxo/token/cache/impl/services/TokenCacheEventProcessorTest.kt]: I'd prefer to see less mocking here. These two classes TokenCache/TokenPoolCache are little more than a wrapper for a map type collection. I'd prefer to use real TokenCache/TokenPoolCache, then we don't have to worry about how accurate our mock behaviour is.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/entities/TokenCacheImpl.kt [hunk 4]
```diff
@@ -1,40 +1,19 @@
...
-    override fun add(tokens: List<CachedToken>) {
-        // This is not very efficient, but it's good enough for this implementation where the list of available tokens
-        // is stored in the state object.
-        val refsToAdd = tokens.map { it.stateRef }.toSet()
-        val existingTokens = tokenPoolCacheState.availableTokens.toMutableList()
-        existingTokens.removeIf { refsToAdd.contains(it.stateRef) }
-        existingTokens.addAll(tokens.map { it.toAvro() })
-        tokenPoolCacheState.availableTokens = existingTokens
+    override fun add(tokens: Collection<CachedToken>) {
+        cachedTokens.putAll(tokens.map { it.stateRef to it })
```
Comment (josephzunigadaly) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/entities/TokenCacheImpl.kt]: This would be a good place to use `associateByTo`

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheComponentFactory.kt [hunk 5]
```diff
@@ -33,18 +40,40 @@ class TokenCacheComponentFactory @Activate constructor(
     private val subscriptionFactory: SubscriptionFactory,
     @Reference(service = ExternalEventResponseFactory::class)
     private val externalEventResponseFactory: ExternalEventResponseFactory,
-    @Reference(service = AvailableTokenService::class)
-    private val availableTokenService: AvailableTokenService
+    @Reference(service = VirtualNodeInfoReadService::class)
+    private val virtualNodeInfoService: VirtualNodeInfoReadService,
```
Comment (josephzunigadaly) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheComponentFactory.kt:44]: Does this belong to this PR? I can see it used but it doesn't look cache related.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/entities/TokenCacheImpl.kt [hunk 6]
```diff
@@ -1,40 +1,19 @@
 package net.corda.ledger.utxo.token.cache.entities
 
-import net.corda.data.ledger.utxo.token.selection.state.TokenPoolCacheState
-import net.corda.ledger.utxo.token.cache.converters.EntityConverter
+class TokenCacheImpl: TokenCache {
```
Comment (josephzunigadaly) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/entities/TokenCacheImpl.kt]: This is inside an exported package. We should put it inside an internal (not OSGi exported) package.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/repositories/impl/UtxoTokenRepositoryImpl.kt [hunk 7]
```diff
@@ -17,27 +11,25 @@ import net.corda.ledger.utxo.token.cache.queries.impl.SqlQueryProviderTokens.Com
...
+import java.math.BigDecimal
+import javax.persistence.EntityManager
+import javax.persistence.Query
+import javax.persistence.Tuple
 
-@Component(service = [UtxoTokenRepository::class])
-class UtxoTokenRepositoryImpl @Activate constructor(
-    @Reference
-    private val sqlQueryProvider: SqlQueryProvider
+class UtxoTokenRepositoryImpl(
```
Comment (josephzunigadaly) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/repositories/impl/UtxoTokenRepositoryImpl.kt:19]: Do the changes to this class belong to the PR? It looks like some unrelated changes made it into the PR.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/services/ServiceConfigurationImpl.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package net.corda.ledger.utxo.token.cache.services
+
+import net.corda.libs.configuration.SmartConfig
+import net.corda.schema.configuration.LedgerConfig.UTXO_TOKEN_CACHED_TOKEN_PAGE_SIZE
+
+class ServiceConfigurationImpl : ServiceConfiguration {
```
Comment (josephzunigadaly) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/services/ServiceConfigurationImpl.kt]: We should move this into an internal (not exported) package.

File: components/ledger/ledger-utxo-token-cache/src/test/kotlin/net/corda/ledger/utxo/token/cache/impl/entities/TokenCacheImplTest.kt [hunk 9]
```diff
@@ -19,62 +18,24 @@ class TokenCacheImplTest {
...
-        val token4 = Token().apply { stateRef = "s4" }
-
-        tokenPoolCacheState.availableTokens = listOf(token1, token2, token3, token4)
-        target.removeAll(setOf("s2", "s4"))
-
-        assertThat(tokenPoolCacheState.availableTokens).containsOnly(token1, token3)
-    }
-
-    @Test
-    fun `iterating the cache returns all tokens from the underlying state object`() {
```
Comment (josephzunigadaly) [inline components/ledger/ledger-utxo-token-cache/src/test/kotlin/net/corda/ledger/utxo/token/cache/impl/entities/TokenCacheImplTest.kt:66]: I think we should keep the replacement and iteration tests.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheComponentFactory.kt (hunk 5 above)
Comment (filipesoliveira) [reply components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheComponentFactory.kt:44]: Yes, this change is intentional.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/repositories/impl/UtxoTokenRepositoryImpl.kt (hunk 7 above)
Comment (filipesoliveira) [reply components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/repositories/impl/UtxoTokenRepositoryImpl.kt:19]: Yes, this is another intentional change.

**Review by josephzunigadaly — DISMISSED** [review]

File: components/ledger/ledger-utxo-token-cache/src/test/kotlin/net/corda/ledger/utxo/token/cache/impl/entities/TokenCacheImplTest.kt (hunk 9 above)
Comment (filipesoliveira) [reply components/ledger/ledger-utxo-token-cache/src/test/kotlin/net/corda/ledger/utxo/token/cache/impl/entities/TokenCacheImplTest.kt:66]: Done.

**Review by josephzunigadaly — APPROVED** [review]
